### PR TITLE
List direct questions for current users or mods

### PR DESCRIPTION
**Pull request by pixeldesu**
Instead of completely removing direct questions from user pages, this now allows moderators (in unmask mode) and users that own the current page to view their direct questions.

### CODE REVIEW

File: app/helpers/user_helper.rb [hunk 1]
```diff
@@ -21,6 +21,10 @@ def moderation_view?
     current_user&.mod? && session[:moderation_view] == true
   end
 
+  def belongs_to_current_user?
```
Comment (nilsding) [inline app/helpers/user_helper.rb]: This method would be better suited as a private method inside UserController; helpers should be for views only.

btw, you can also use Ruby3's endless method definition syntax:
```ruby
private

def belongs_to_current_user? = @user == current_user
```

**Review by nilsding — CHANGES_REQUESTED** [review]

**Review by nilsding — APPROVED** [review]

**Comment by pixeldesu** [comment]
So I adjusted this, but I didn't manage to properly get tests done for `UserController#questions`. Here's my attempt in a gist: https://gist.github.com/pixeldesu/f80868e0c196227c0adb3e6780ae31b9

The first test was similar to the others but that didn't work either and checking for the size of the instance variable doesn't yield the right answer too, even though it works fine as expected in the frontend. (the checks for no renders/null length pass because the first one fails, etc. etc.)

**Comment by nilsding** [comment]
it's because you used `let` instead of `let!`; `let!` expressions are evaluated eagerly whereas `let` ones will get evaluated the first time they're called

**Comment by pixeldesu** [comment]
It wasn't working with a question being created in a `before` block either. And I just tried it, `@question.size` still is 0 with `let!`.

**Comment by nilsding** [comment]
Ah, found it.  The question is created as an anonymous one.  Pass `author_is_anonymous: false` to the `FactoryBot.create` call and then it does what it should.

**Comment by pixeldesu** [comment]
Yup, that fixed that.

I will adjust the other tests (the moderation one wasn't properly implemented yet) and then push those. Wanted to have this behaviour tested properly because it could lead to unwanted infos being leaked.

File: spec/controllers/user_controller_spec.rb [hunk 2]
```diff
@@ -50,15 +52,57 @@
...
+    let!(:question) { FactoryBot.create(:question, user:, direct: true, author_is_anonymous: false) }
     subject { get :questions, params: { username: user.screen_name } }
 
+    it "renders the user/questions template" do
+      subject
+      expect(assigns(:user)).to eq(user)
+      expect(response).to render_template("user/questions")
+    end
+
+    context "current user signed in" do
```
Comment (nilsding) [inline spec/controllers/user_controller_spec.rb]:
```suggestion
    context "when the current user is the question author" do
```

File: spec/controllers/user_controller_spec.rb [hunk 3]
```diff
@@ -50,15 +52,57 @@
...
+        sign_in user
+      end
+
+      it "renders all questions" do
+        subject
+        expect(assigns(:questions).size).to eq(1)
+      end
+    end
+
     context "user signed in" do
```
Comment (nilsding) [inline spec/controllers/user_controller_spec.rb]:
```suggestion
    context "when the current user is someone else" do
```

File: spec/controllers/user_controller_spec.rb [hunk 4]
```diff
@@ -50,15 +52,57 @@
...
-      it "renders the user/questions template" do
+      it "renders no questions" do
         subject
-        expect(assigns(:user)).to eq(user)
-        expect(response).to render_template("user/questions")
+        expect(assigns(:questions).size).to eq(0)
+      end
+    end
+
+    context "moderator unmasked" do
```
Comment (nilsding) [inline spec/controllers/user_controller_spec.rb]:
```suggestion
    context "when a moderator uses unmask" do
```

File: spec/controllers/user_controller_spec.rb [hunk 5]
```diff
@@ -50,15 +52,57 @@
...
+        allow_any_instance_of(UserHelper).to receive(:moderation_view?) { true }
+      end
+
+      it "contains all questions" do
+        subject
+        expect(assigns(:questions).size).to eq(1)
+      end
+    end
+
+    context "user not signed in" do
```
Comment (nilsding) [inline spec/controllers/user_controller_spec.rb]:
```suggestion
    context "when user is not signed in" do
```

**Review by nilsding — APPROVED** [review]

**Review by nilsding — APPROVED** [review]

**Pull request merged**